File: src/main/java/net/guizhanss/guizhancraft/core/services/LocalizationService.java
package net.guizhanss.guizhancraft.core.services;

import java.text.MessageFormat;
import java.util.Locale;

import javax.annotation.Nonnull;
import javax.annotation.ParametersAreNonnullByDefault;

import com.google.common.base.Preconditions;

import net.guizhanss.guizhanlib.minecraft.utils.ChatUtil;

import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;

import net.guizhanss.guizhancraft.implementation.GuizhanCraft;
import net.guizhanss.guizhanlib.localization.Localization;
import net.guizhanss.guizhanlib.utils.StringUtil;
import net.md_5.bungee.api.ChatMessageType;
import net.md_5.bungee.api.chat.BaseComponent;
import net.md_5.bungee.api.chat.TextComponent;

/**
 * This is an extended {@link Localization} class.
 *
 * @author ybw0014
 */
@SuppressWarnings("ConstantConditions")
public final class LocalizationService extends Localization {
    public LocalizationService(GuizhanCraft plugin) {
        super(plugin);
    }

    @ParametersAreNonnullByDefault
    @Nonnull
    public String getString(String key, Object... args) {
        return MessageFormat.format(getString(key), args);
    }

    @Nonnull
    public String getCategoryName(@Nonnull String categoryId) {
        Preconditions.checkArgument(categoryId != null, "Category id cannot be null");

        return getString("categories." + StringUtil.dehumanize(categoryId).toLowerCase(Locale.ROOT));
    }

    @Nonnull
    public String getItemName(@Nonnull String itemId) {
        Preconditions.checkArgument(itemId != null, "Item id cannot be null");

        return getString("items." + StringUtil.dehumanize(itemId).toLowerCase(Locale.ROOT) + ".name");
    }

    @Nonnull
    public String[] getItemLore(@Nonnull String itemId) {
        Preconditions.checkArgument(itemId != null, "Item id cannot be null");

        return getStringArray("items." + StringUtil.dehumanize(itemId).toLowerCase(Locale.ROOT) + ".lore");
    }

    @Nonnull
    public String getResearchName(@Nonnull String researchId) {
        Preconditions.checkArgument(researchId != null, "Research Id cannot be null");

        return getString("researches." + StringUtil.dehumanize(researchId).toLowerCase(Locale.ROOT));
    }

    @ParametersAreNonnullByDefault
    public <T extends CommandSender> void send(T sender, String key, Object... args) {
        Preconditions.checkArgument(sender != null, "sender cannot be null");
        Preconditions.checkArgument(key != null, "Key cannot be null");

        ChatUtil.send(sender, MessageFormat.format(getString(key), args));
    }

    @ParametersAreNonnullByDefault
    public <T extends CommandSender> void sendMessage(T sender, String messageKey, Object... args) {
        Preconditions.checkArgument(messageKey != null, "Message key cannot be null");

        send(sender, "messages." + messageKey, args);
    }

    @ParametersAreNonnullByDefault
    public void sendConsole(String messageKey, Object... args) {
        Preconditions.checkArgument(messageKey != null, "Message key cannot be null");

        GuizhanCraft.sendConsole(getString("console." + messageKey), args);
    }

    @ParametersAreNonnullByDefault
    public void sendActionbarMessage(Player p, String messageKey, Object... args) {
        Preconditions.checkArgument(p != null, "Player cannot be null");
        Preconditions.checkArgument(messageKey != null, "Message key cannot be null");

        String message = MessageFormat.format(getString("messages." + messageKey), args);

        BaseComponent[] components = TextComponent.fromLegacyText(ChatUtil.color(message));
        p.spigot().sendMessage(ChatMessageType.ACTION_BAR, components);
    }
}

File: src/main/kotlin/net/guizhanss/guizhancraft/core/services/LocalizationService.kt
package net.guizhanss.guizhancraft.core.services

import net.guizhanss.guizhancraft.GuizhanCraft
import net.guizhanss.guizhancraft.utils.listYmlFilesInJar
import net.guizhanss.guizhancraft.utils.toId
import net.guizhanss.guizhanlib.kt.slimefun.items.builder.MaterialType
import net.guizhanss.guizhanlib.kt.slimefun.items.toItem
import net.guizhanss.guizhanlib.minecraft.utils.ChatUtil
import net.guizhanss.guizhanlib.slimefun.addon.SlimefunLocalization
import org.bukkit.command.CommandSender
import org.bukkit.inventory.ItemStack
import java.io.File
import java.text.MessageFormat

class LocalizationService(
    private val plugin: GuizhanCraft,
    private val jarFile: File
) : SlimefunLocalization(plugin) {

    init {
        extractTranslations()
    }

    private fun extractTranslations() {
        val translationsFolder = File(plugin.dataFolder, FOLDER_NAME)
        if (!translationsFolder.exists()) {
            translationsFolder.mkdirs()
        }
        val translationFiles = listYmlFilesInJar(jarFile, FOLDER_NAME)
        for (translationFile in translationFiles) {
            val filePath = FOLDER_NAME + File.separator + translationFile
            plugin.saveResource(filePath, true)
        }
    }

    fun getString(key: String, vararg args: Any?): String = MessageFormat.format(getString(key), *args)

    // categories
    fun getItemGroupItem(key: String, materialType: MaterialType): ItemStack =
        getItem("GROUP_${key.toId()}", materialType.convert()).toItem()

    // recipe
    fun getRecipeTypeItem(key: String, materialType: MaterialType): ItemStack =
        getItem("RECIPE_${key.toId()}", materialType.convert()).toItem()

    // ui
    fun getUIItem(key: String, materialType: MaterialType): ItemStack =
        getItem("UI_${key.toId()}", materialType.convert()).toItem()

    // items
    fun getItemName(itemId: String, vararg args: Any?) = getString("items.${itemId.toId()}.name", *args)
    fun getItemLore(itemId: String): List<String> = getStringList("items.${itemId.toId()}.lore")

    fun sendMessage(sender: CommandSender, key: String, vararg args: Any) {
        if (GuizhanCraft.integrationService.slimefunTranslationEnabled) {
            GuizhanCraft.integrationService.sendMessage(sender, key, *args)
        } else {
            ChatUtil.send(sender, MessageFormat.format(getString("messages.$key"), *args))
        }
    }

    companion object {

        const val FOLDER_NAME = "lang"
    }
}
